### PR TITLE
qa: Add --filter option to test_runner.py

### PR DESCRIPTION
**Pull request by promag**
Allows to run functional tests like:
```sh
test/functional/test_runner.py --filter wallet
```

### CODE REVIEW

File: test/functional/test_runner.py [hunk 1]
```diff
@@ -234,6 +234,7 @@ def main():
     parser.add_argument('--quiet', '-q', action='store_true', help='only print dots, results summary and failure logs')
     parser.add_argument('--tmpdirprefix', '-t', default=tempfile.gettempdir(), help="Root directory for datadirs")
     parser.add_argument('--failfast', action='store_true', help='stop execution after the first test failure')
+    parser.add_argument('--filter', help='regular expression to filter scripts to run')
```
Comment (jonatack) [inline test/functional/test_runner.py]: Perhaps place it alphabetically, e.g. between extended and help. The failfast placement looks like an anomaly.

Suggested help: "filter tests to run by regular expression", or possibly "filter scripts to run by regular expression", in the format of verb + subject + details. FWIW I've rewritten the other argument help messages here and in test_framework.py for a PR following this format.

File: test/functional/test_runner.py [hunk 2]
```diff
@@ -294,6 +295,9 @@ def main():
             if not exclude_list:
                 print("{}WARNING!{} Test '{}' not found in current test list.".format(BOLD[1], BOLD[0], exclude_test))
 
+    if args.filter:
+        test_list = list(filter(re.compile(args.filter).search, test_list))
```
Comment (jonatack) [inline test/functional/test_runner.py]: Tested successfully with one exception: passing "tool" as in `test_runner.py --filter tool` or `test/functional/test_runner.py --filter tool` does no filtering and runs all the tests.

**Review by jonatack — COMMENTED** [review]
Concept ACK / tested ACK

File: test/functional/test_runner.py (hunk 1 above)
Comment (promag) [reply test/functional/test_runner.py]: Argument `--ci` is also misplaced. Note that `test/functional/test_runner.py --help` already has the output sorted.

Updated message.

Comment (jonatack) [reply test/functional/test_runner.py]: When I run `test/functional/test_runner.py --help` on Linux Debian, the output is not sorted (it's a bit of a mess TBH and am writing a PR to improve that and a few other things with the test runner output).

**Comment by jonatack** [comment]
ACK 1a6242526093424947eb49f3416dc0c6bc9fc3a8

**Comment by jnewbery** [comment]
Just saw this. What's wrong with `./test_runner.py wallet*`?

**Comment by promag** [comment]
Nothing wrong, you have to `cd test/functional/` or something like that. This is also similar to `bench_bitcoin -filter`.

**Comment by jonatack** [comment]
There is no further need to cd into test/functional/ with #16374. Maybe I should have been more explicit in my review; I saw this PR as an optional complement to that PR but not a replacement.

**Comment by MarcoFalke** [comment]
See the IRC discussion:

> Currently, passing wildcard testname args to the test runner from outside the test/functional/ directory does not work. See these recent IRC discussions for more background: http://www.erisian.com.au/bitcoin-core-dev/log-2019-07-10.html#l-262 (lines 262 to 323) and http://www.erisian.com.au/bitcoin-core-dev/log-2019-07-11.html#l-134.

From https://github.com/bitcoin/bitcoin/pull/16374#issue-296813072

**Pull request merged**